oscarapi/middleware.py

```diff
@@ -0,0 +1,167 @@
+    and non trusted clients, see README.rst.
+    """
+
+    def process_request(self, request):
+        """
+        Parse the session id from the 'Session-Id: ' header when using the api.
+        """
+        if is_api_request(request):
+            try:
+                parsed_session_uri = parse_session_id(request)
+                if parsed_session_uri is not None:
+                    domain = get_domain(request)
+                    if parsed_session_uri['realm'] != domain:
+                        raise exceptions.NotAcceptable(
+                            _('Can not accept cookie with realm %s on realm %s') % (
+                                parsed_session_uri['realm'],
+                                domain
+                            )
+                        )
+                    session_id = session_id_from_parsed_session_uri(
+                        parsed_session_uri)
+                    request.session = start_or_resume(
+                        session_id, session_type=parsed_session_uri['type'])
+                    request.parsed_session_uri = parsed_session_uri
+
+                    # since the session id is assigned by the CLIENT, there is
+                    # no point in having csrf_protection. Session id's read
+                    # from cookies, still need csrf!
+                    request.csrf_processing_done = True
+                    return None
+            except exceptions.APIException as e:
+                response = HttpResponse('{"reason": "%s"}' % e.detail)
+                response.status_code = e.status_code
+                return response
+
+        return super(HeaderSessionMiddleware, self).process_request(request)
+
+    def process_response(self, request, response):
+        """
+        Add the 'Session-Id: ' header when using the api.
+        """
+        if is_api_request(request) \
+                and getattr(request, 'session', None) is not None \
+                and hasattr(request, 'parsed_session_uri'):
+            session_key = request.session.session_key
+            parsed_session_key = session_id_from_parsed_session_uri(
+                request.parsed_session_uri)
+            assert(session_key == parsed_session_key), \
+                    '%s is not equal to %s' % (session_key, parsed_session_key)
+            response['Session-Id'] = \
+                'SID:%(type)s:%(realm)s:%(session_id)s' % (
+                    request.parsed_session_uri)
+
+        return super(HeaderSessionMiddleware, self).process_response(
+            request, response)
+
+
+class ApiGatewayMiddleWare(object):
+    """
+    Protect the api gateway with token.
+    """
+    def process_request(self, request):
+        if is_api_request(request):
+            key = authentication.get_authorization_header(request)
+            if models.ApiKey.objects.filter(key=key).exists():
+                return None
+
+            logger.error('No credentials provided')
```

Comment:
This log message would be more useful with some additional info about the request like the client IP address, the requested path etc.